File: day15.py
"""Solution to day 15 of Advent of Code"""

from get_input import get_input
from itertools import count, islice


def game(starting):
    turns = {n: t for t, n in enumerate(starting[:-1], 1)}
    yield from iter(starting)
    last = starting[-1]
    for turn in count(len(starting)):
        if last not in turns:
            nxt = 0
        else:
            nxt = turn - turns[last]
        turns[last] = turn
        yield nxt
        last = nxt


def part1(lines):
    return next(islice(game(lines), 2020-1, None))


def part2(lines):
    return next(islice(game(lines), 30000000-1, None))


def parse(line):
    return tuple(int(n) for n in line.strip().split(','))


if __name__ == "__main__":
    assert part1((0, 3, 6)) == 436
    assert part1((1, 3, 2)) == 1
    assert part1((2, 1, 3)) == 10
    assert part1((1, 2, 3)) == 27
    assert part1((3, 1, 2)) == 1836
    assert part1((3, 2, 1)) == 438
    assert part1((3, 2, 1)) == 438
    LINES = parse(get_input(day=15, year=2020))
    print(f"Part 1: {part1(LINES)}")
    print(f"Part 2: {part2(LINES)}")

File: day15.rs
use common::get_input;
use std::collections::HashMap;

type Input = Vec<usize>;

struct Game<'a> {
    turns: HashMap<usize, usize>,
    starting: &'a[usize],
    turn: usize,
    last: Option<usize>,
}

impl<'a> Game <'a> {
    fn new(starting: &'a[usize] ) -> Game {
        Game {
            turns: HashMap::new(),
            starting,
            turn: 0,
            last: None,
        }
    }
}

impl<'a> Iterator for Game<'a> {
    type Item = usize;

    fn next(&mut self) -> Option<usize> {
        self.turn += 1;
        let next = if let Some(&initial) = self.starting.get(self.turn-1) {
            initial
        } else if let Some(last) = self.turns.get(&self.last.unwrap()) {
            self.turn - last
        } else {
            0
        };
        // println!("{}: {} - {:?}", self.turn, next, self.turns);
        if let Some(last) = self.last {
            self.turns.insert(last, self.turn);
        }
        self.last = Some(next);
        Some(next)
    }
}

fn part1(input: &Input) -> usize {
    Game::new(input).nth(2020-1).unwrap()
}

fn part2(input: &Input) -> usize {
    Game::new(input).nth(30000000-1).unwrap()
}

fn parse(text: String) -> Input {
    text.trim()
        .split('\n')
        .map(|line| line.split(',').map(|n| n.parse::<usize>().unwrap()).collect())
        .next()
        .unwrap()
}

fn main() {
    assert_eq!(part1(&vec![0, 3, 6]), 436);
    assert_eq!(part1(&vec![1, 3, 2]), 1);
    assert_eq!(part1(&vec![2, 1, 3]), 10);
    assert_eq!(part1(&vec![1, 2, 3]), 27);
    assert_eq!(part1(&vec![3, 1, 2]), 1836);
    assert_eq!(part1(&vec![3, 2, 1]), 438);
    let input = parse(get_input(15, 2020));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
